200 label uprns in city centre

eglucas opened the pull request:
Fixes https://github.com/nestauk/asf_heat_pump_suitability/issues/200

# Description

Code to process UPRNs and identify whether they are in a potential heat network zone, based on which spatial signature type area they are in.

Changed files:
- `config/base.yaml` - add path to datasets containing the spatial signatures dataset for GB. There is one version with the suffix `_simplified` and the other without a suffix. **The code and run script I've developed is based only on the `_simplified` version. (I've not been able to inspect the full version as loading it keeps crashing my kernel 🥲)**

- `getters/load_geodata.py` - add getter for spatial signatures dataset for GB with arguments to select `simplified` or `full` version.

- `pipeline/transform/heat_network_zones.py` - add function to label UPRNs that are in one of the given spatial signature types. This function does the same processing as the HNZ labeller, so there might be opportunity to streamline but I think keeping them distinctly separate is best for clarity at this stage.
Note: I've changed the logic for `usecols` in both the HNZ and spatial signature labeller so that default `usecols=None` no longer means that all columns are used but rather only the columns essential for processing.

- `pipeline/run/heat_network_city_centres.py` - add code block to existing script that adds boolean flags labelling residential UPRNs indicating whether they're in one of spatial signature types we're currently considering to be representative of a city centre. This script now saves two separate dataframes to S3 - one with existing HNZ labels, and one with spatial signature types and city centre boolean labels. For simplicity, I've kept them separate but we could in future, combine both HNZ and city centre labels one dataframe by converting `hp_zone_uprn_df` to a `gdf` and passing that to `heat_network_zones.label_gdf_heat_network_spatial_signature_uprns()`.

- `analysis/spatial_signatures_hn_zones.py` - add analytical notebook that compares existing HNZ and city centre spatial signature types. This is specific to Plymouth LA. Open as Jupyter Notebook.

# Instructions for Reviewer

To run the script: python asf_heat_pump_suitability/pipeline/run/heat_network_city_centres.py --local_authorities plymouth
- Please inspect the new `city_centre_uprn_df` that is saved out to S3

Please pay special attention to
* Any bugs or opportunities for improvement in the loading and labelling functions, as well as the new code block in the run script
* Whether you're happy with the saved location of the spatial signatures dataset, as well as how their paths are inserted in config
* Schema of the saved output dataframe

With regards to the analysis notebook, please run all the cells and have a look at the outputs. Let me know if you have any other thoughts on how we're evaluating whether the city centre signature types do a good enough job at flagging potential heat network zones. Please also flag if there are any errors in the analysis or interpretation.

# Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [ ] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [x] I have explained this PR above
- [x] I have requested a code review

eglucas committed:
add spatial signature data getter

eglucas committed:
add spatial signature label function

eglucas committed:
add city centre identifier to run script

eglucas committed:
add coverage analysis notebook

crispy-wonton commented on asf_heat_pump_suitability/getters/load_geodata.py:
```diff
@@ -57,3 +57,39 @@ def load_gdf_heat_network_zones(local_authority: str, **kwargs) -> gpd.GeoDataFr
+
+    Args:
+        detail_level (str, optional): Which level of descriptive detail to load.
+            Must be either "simplified" or "full". Defaults to "simplified".
```
Does `simplified` just mean fewer columns loaded? If so, I would change the word to something else here. 'simplified' already has a meaning when talking about GeoDataFrames and refers to a simplification of the geometries themselves, e.g. reducing the number of points in a polygon.

You could change it to a `boolean` argument and call it something like `required_fields` or whatever you think is intuitive and then have it load either the full dataset or only the required fields.

eglucas replied:
Thank you so much for raising this - I didn't know about this "simplified" convention with geodataframes. As we've discussed on slack, I compared the number of polygon coordinate points for a chosen geometry in the simplified vs. full-res and there was 264 in simple and 720 in full. This does mean that the simplified suffix in the file name is referring to the simplified polygon geometry.

I have rewritten the getter docstring to reflect this and changed the default to be `detail_level = "full"`

crispy-wonton commented on asf_heat_pump_suitability/getters/load_geodata.py:
```diff
@@ -57,3 +57,39 @@ def load_gdf_heat_network_zones(local_authority: str, **kwargs) -> gpd.GeoDataFr
+) -> gpd.GeoDataFrame:
+    """
+    Load GeoDataFrame with polygons in GB from the Spatial Signatures Framework  classified by their
+    Spatial Signature type. (Source: https://doi.org/10.6084/m9.figshare.16691575).
```
Please can you add the crs to the description here :)

crispy-wonton commented on asf_heat_pump_suitability/pipeline/transform/heat_network_zones.py:
```diff
@@ -54,3 +57,64 @@ def label_gdf_heat_network_zone_uprns(
+        spatial_signatures_gdf (gpd.GeoDataFrame): polygons of spatial signatures
+        usecols (list, optional): names of descriptive columns from spatial_signatures_gdf to be joined with uprn_gdf (in addition to "geometry" and "type").
+            Default is None to use only "geometry" and "type".
+        types (list, optional): spatial signature types potentially suitable for a heat network. Default is None to use all spatial signature types.
```
I would probably set the default spatial signature types to those you've selected. You could put them as a global variable at the top of this file. Otherwise, here I think we could get another case of silent fail where everything gets labelled as potentially suitable for a HNZ and we can't immediately see why.

crispy-wonton commented on asf_heat_pump_suitability/pipeline/transform/heat_network_zones.py:
```diff
@@ -54,3 +57,64 @@ def label_gdf_heat_network_zone_uprns(
+    Args:
+        uprn_gdf (gpd.GeoDataFrame): UPRNs with point geometries to be labelled
+        spatial_signatures_gdf (gpd.GeoDataFrame): polygons of spatial signatures
+        usecols (list, optional): names of descriptive columns from spatial_signatures_gdf to be joined with uprn_gdf (in addition to "geometry" and "type").
```
Is there a reason you allowed this functionality instead of just going with the default options you've selected? If not, we can cut this out to simplify the function.

eglucas replied:
It was initially because I didn't know what other columns were in the full-res version of the dataset. Now that I've been able to load just a chunk of it and inspected the column names - I agree that it's best to remove this functionality.

crispy-wonton commented on asf_heat_pump_suitability/pipeline/transform/heat_network_zones.py:
```diff
@@ -54,3 +57,64 @@ def label_gdf_heat_network_zone_uprns(
+    else:
+        spatial_signatures_gdf = spatial_signatures_gdf[["geometry", "type"]]
+
+    # Spatial join for for labelling UPRN with signature type
```
```suggestion
    # Spatial join for labelling UPRN with signature type
```

crispy-wonton commented on asf_heat_pump_suitability/pipeline/transform/heat_network_zones.py:
```diff
@@ -54,3 +57,64 @@ def label_gdf_heat_network_zone_uprns(
+    labelled_uprn_gdf = uprn_gdf.sjoin(
+        spatial_signatures_gdf,
+        how="left",
+        predicate="intersects",  # include properties intersecting spatial signature cell boundary
```
We have to be careful that we don't end up with UPRNs (if not in Plymouth, then it could happen with scaling) with multiple spatial signatures here in cases when a UPRN is on a border because of using the `intersects` predicate. This could mean you end up with UPRNs which are labelled as both being within and not within a potential HN zone.

I think the solution would be to keep the predicate at `intersects`, but set a rule so that any UPRNs which end up in this situation get assigned `in_potential_hn_zone == True`. 

I've checked the `s3://asf-heat-pump-suitability/local_heat_planning/outputs/plymouth_residential_uprns_with_city_centres.parquet` output and there are no duplicated UPRNs so it doesn't seem to be a problem at the moment but let's put some code in place to address it in case these edge cases exist when scaling.

eglucas replied:
Thanks so much for raising this important point that I hadn't considered!

I've ended up implementing (in https://github.com/nestauk/asf_heat_pump_suitability/pull/203/commits/04316c7d3a77dc46fc2615b0c1d2a04a728b3095) what I think is pretty clean:
- all matched spatial signature types for each UPRN will be combined into a list
- `in_city_centre=True` for that UPRN if any type in the list is in `CITY_CENTRE_TYPES`

This way also means that all matched spatial signatures are preserved if they need to be inspected later.

crispy-wonton replied:
Great idea to retain them in a list :)

crispy-wonton commented on asf_heat_pump_suitability/pipeline/transform/heat_network_zones.py:
```diff
@@ -54,3 +57,64 @@ def label_gdf_heat_network_zone_uprns(
+    ).drop(columns="index_right")
+
+    # Add potential heat network zone boolean label to original uprn_gdf
+    labelled_uprn_gdf["in_potential_hn_zone"] = (
```
I would relabel this as `in_city_centre`. I think this label has the potential to cause some confusion given that DESNZ's zones could be thought of as 'potential' HN zones.

I know that `in_city_centre` also isn't perfectly accurate but since it aligns with our phrasing in the decision tree I think that's clearer for now. Open to other ideas.

sofiapinto replied:
i'd use `in_city_centre`, like in the decision tree. otherwise it's confusing.

crispy-wonton commented on asf_heat_pump_suitability/pipeline/transform/heat_network_zones.py:
```diff
@@ -54,3 +57,64 @@ def label_gdf_heat_network_zone_uprns(
+    )
+    labelled_uprn_gdf["in_potential_hn_zone"] = labelled_uprn_gdf[
+        "in_potential_hn_zone"
+    ].fillna(False)
```
Why would there be NA values in this column? Is it because some UPRNs do not get assigned a spatial signature?

sofiapinto replied:
same as above

crispy-wonton commented on asf_heat_pump_suitability/pipeline/transform/heat_network_zones.py:
```diff
@@ -54,3 +57,64 @@ def label_gdf_heat_network_zone_uprns(
+    # Return as polars df without geometry
+    labelled_uprn_df = pl.from_pandas(labelled_uprn_gdf.drop(columns="geometry"))
+
+    labelled_uprn_df = labelled_uprn_df.rename({"type": "spatial_signature_type"})
```
```suggestion
    labelled_uprn_df = pl.from_pandas(labelled_uprn_gdf.drop(columns="geometry")).rename({"type": "spatial_signature_type"})
```

crispy-wonton commented on asf_heat_pump_suitability/pipeline/run/heat_network_city_centres.py:
```diff
@@ -90,8 +130,14 @@ def parse_arguments() -> argparse.Namespace:
         raise NotImplementedError("Processing for all of GB is not yet supported.")
 
-    # Save residential UPRNs with labels to S3
+    # Save residential UPRNs with existing hn zone labels to S3
```
I think we can save HNZ and city centre labels into the same output file - lmk if there's a reason you kept separate :)

eglucas replied:
Agree! Will combine into one output file

sofiapinto commented on asf_heat_pump_suitability/config/base.yaml:
We need to also update this https://github.com/nestauk/asf_heat_pump_suitability/blob/dev/asf_heat_pump_suitability/config/README.md either in this or in a separate PR with the two new datasets, the HNZ doc and the spatial signatures framework.

eglucas replied:
Thanks Sofia, I've created a specific issue for this: https://github.com/nestauk/asf_heat_pump_suitability/issues/205

sofiapinto commented on asf_heat_pump_suitability/getters/load_geodata.py:
```diff
@@ -57,3 +57,39 @@ def load_gdf_heat_network_zones(local_authority: str, **kwargs) -> gpd.GeoDataFr
+    )
+
+    print(
+        f"Spatial signatures {detail_level} geodataframe successfully loaded with CRS {gdf.crs}."
```
Nice! I love that we print the CRS too, so useful.

sofiapinto commented on asf_heat_pump_suitability/pipeline/run/heat_network_city_centres.py:
```diff
@@ -55,13 +55,15 @@ def parse_arguments() -> argparse.Namespace:
         )
 
-        # Filter for UPRNs in existing heat network zones
+        # Label UPRNs in existing heat network zones
```
I wonder if it might be useful to clarify
```suggestion
        # Label UPRNs in existing, potential and planned heat network zones
```

sofiapinto commented on asf_heat_pump_suitability/pipeline/run/heat_network_city_centres.py:
```diff
@@ -78,7 +80,45 @@ def parse_arguments() -> argparse.Namespace:
+            "Regional urbanity",
+            "Local urbanity",
+            "Dense urban neighbourhoods",
+        ]
```
i'd check with @crispy-wonton about how she's been doing this across the repo, but i'd suggest we maybe want to move this to `base.yaml`?

sofiapinto replied:
I think the areas selected sound good for our MVP!

eglucas replied:
I've ended up implementing @crispy-wonton 's suggestion to set this list of types as a global variable `CITY_CENTRE_TYPES` at the top of new `transform/city_centres.py` (and is now the default argument for `types`)

sofiapinto commented on asf_heat_pump_suitability/pipeline/run/heat_network_city_centres.py:
```diff
@@ -78,7 +80,45 @@ def parse_arguments() -> argparse.Namespace:
+                "X_COORDINATE",
+                "Y_COORDINATE",
+                "spatial_signature_type",
+                "in_potential_hn_zone",
```
i''d call this one `in_city_centre` in the original function

sofiapinto commented on asf_heat_pump_suitability/pipeline/run/heat_network_city_centres.py:
```diff
@@ -90,8 +130,14 @@ def parse_arguments() -> argparse.Namespace:
     )
+
+    # Save residential UPRNs with city centre labels to S3
+    save_utils.save_to_s3(
+        city_centre_uprn_df,
+        f"s3://asf-heat-pump-suitability/local_heat_planning/outputs/{args.local_authorities}_residential_uprns_with_city_centres.parquet",
+    )
```
I'll leave it to @crispy-wonton on how data should be saved but perhaps these two flags could be saved in the same file instead of separate files? Two separate columns, but same file.

We also have a `plymouth_residential_uprns_with_features.parquet` where eventually all relevant columns will be saved, I imagine. I'm guessing that is a product of Roisin's work in a separate PR.

eglucas replied:
Thanks Sofia, yes - Roisin made the same suggestion. Will combine into one output file!

crispy-wonton commented on asf_heat_pump_suitability/research/exploratory/20251205_plymouth_investigation.py:
```diff
@@ -0,0 +1,530 @@
+
+# %%
+# Confirming spatial signature types
+print("Spatial signatures:")
+counter = 1
+for type in spatial_signatures_gb_simplified_gdf["type"].unique():
+    print(f"{counter}. {type}")
+    counter += 1
```
I love this! 😄 the counter delighted me

crispy-wonton commented on asf_heat_pump_suitability/research/exploratory/20251205_plymouth_investigation.py:
```diff
@@ -0,0 +1,530 @@
+
+fig, ax = plt.subplots(figsize=(15, 7.5))
+
+# plot spatial signatures polygons with mapped colors
```
nice plot!

crispy-wonton commented on asf_heat_pump_suitability/research/exploratory/20251205_plymouth_investigation.py:
```diff
@@ -0,0 +1,530 @@
+def inspect_spatial_signatures_in_hn_zone(zone_id: str):
+    zone_gdf = plymouth_hn_zones_gdf[plymouth_hn_zones_gdf["ZoneID"] == zone_id]
+
+    just_text_raw = str(zone_gdf.Just.values[0])
+    just_text = "\n".join(textwrap.wrap(just_text_raw, width=90))
```
```suggestion
    # Add HN zone notes 
    just_text_raw = str(zone_gdf.Just.values[0])
    just_text = "\n".join(textwrap.wrap(just_text_raw, width=90))
```
Something like this, because it's not clear from looking at it what `Just` is.

crispy-wonton replied:
Nice! haven't seen `textwrap` before - looks very useful

crispy-wonton commented on asf_heat_pump_suitability/analysis/spatial_signatures_hn_zones/20251205_plymouth_investigation.py:
```diff
@@ -0,0 +1,530 @@
+    edgecolor="#1f77b4",
+    linewidth=2,
+    alpha=0.8,
+    label="HNZ Zones",
```
```suggestion
    label="HN Zones",
```

crispy-wonton commented on asf_heat_pump_suitability/research/exploratory/20251205_plymouth_investigation.py:
```diff
@@ -0,0 +1,530 @@
+)
+
+# %% [markdown]
+# **Comparison of zones by UPRNs coverage**
```
You could also plot a confusion matrix here: https://scikit-learn.org/stable/modules/generated/sklearn.metrics.ConfusionMatrixDisplay.html

crispy-wonton commented on asf_heat_pump_suitability/analysis/spatial_signatures_hn_zones/20251205_plymouth_investigation.py:
```diff
@@ -0,0 +1,530 @@
+# What % of residential UPRNs are in existing HNZ?
+prop_in_hn_zone = plymouth_uprns_hnz_labelled_df.select(
+    pl.col("in_hn_zone").mean()
+).item()
```
```suggestion
prop_in_hn_zone = plymouth_uprns_hnz_labelled_df["in_hn_zone"].mean()
```

crispy-wonton commented on asf_heat_pump_suitability/research/exploratory/20251205_plymouth_investigation.py:
```diff
@@ -0,0 +1,530 @@
+
+summary_df
+
+# %% [markdown]
```
This is a great summary and much easier to understand than the table above. Especially with the 236% value which I'm not sure how to interpret..

crispy-wonton commented on asf_heat_pump_suitability/research/exploratory/20251205_plymouth_investigation.py:
```diff
@@ -0,0 +1,530 @@
+# Inspecting existing HNZ one at a time:
+
+# %%
+plymouth_hn_zones_with_labels_gdf = gpd.sjoin(
+    plymouth_hn_zones_gdf,
+    spatial_signatures_plymouth_gdf[["type", "geometry"]],
+    how="left",
+    predicate="within",
+)
+
+plymouth_hn_zones_with_labels_agg_gdf = (
+    plymouth_hn_zones_with_labels_gdf.groupby("ZoneID")["type"]
+    .apply(lambda x: list(x.dropna().unique()))
+    .reset_index()
+)
```
Is this meant to be deleted?

eglucas replied:
No, I've just not cleaned up the way that the helper function `inspect_spatial_signatures_in_hn_zone` is used. I need to create `plymouth_hn_zones_with_labels_agg_gdf` which is then looped over in the following code in subsequent lines

`for zone_id in plymouth_hn_zones_with_labels_agg_gdf["ZoneID"].unique():
    inspect_spatial_signatures_in_hn_zone(zone_id)`

This section can definitely be cleaned up but given that it's just for the purpose of this notebook and visualising specific outputs for exploration, I've left it as is.

crispy-wonton commented on asf_heat_pump_suitability/analysis/spatial_signatures_hn_zones/20251205_plymouth_investigation.py:
```diff
@@ -0,0 +1,530 @@
+hnz_true_df = plymouth_uprns_hnz_spatial_signature_labelled_df.filter(
+    (pl.col("in_hn_zone") == True)
+)
+prop_true = hnz_true_df.select(pl.col("in_potential_hn_zone").mean()).item()
```
```suggestion
prop_true = hnz_true_df["in_potential_hn_zone"].mean()
```

crispy-wonton commented on asf_heat_pump_suitability/analysis/spatial_signatures_hn_zones/20251205_plymouth_investigation.py:
```diff
@@ -0,0 +1,530 @@
+            f"{(overlap_area / area_hnz) * 100:.1f}%",
+        ],
+        "% of selected signatures covered by HNZ": [
+            f"{(area_hnz / area_selected) * 100:.1f}%",
```
This is just a size comparison right because there's no overlay involved and doesn't actually calculate percentage coverage?

eglucas replied:
Ah good spot! This is a mistake. I've now corrected it to:
```suggestion
            f"{percent_hnz_covered_by_selected:.1f}%",
```

I've also cleaned up the summary table so it's clearer.

crispy-wonton commented on asf_heat_pump_suitability/analysis/spatial_signatures_hn_zones/20251205_plymouth_investigation.py:
```diff
@@ -0,0 +1,530 @@
+)
+
+# %% [markdown]
+# 05/12/25: City centres as a proxy performs better at capturing properties in existing HNZ, compared to land area. Mainly because of the higher density of properties in the selected urban area types.
```
What do you mean 'compared to land area'?

eglucas replied:
I meant spatial coverage - have corrected this text.

```suggestion
# 05/12/25: City centres as a potential proxy performs better at capturing properties in existing HNZ (i.e. UPRN coverage), compared to spatial coverage. Mainly because of the higher density of properties in the selected urban area types.
```

crispy-wonton commented on asf_heat_pump_suitability/research/exploratory/20251205_plymouth_investigation.py:
Please could you move this file to `asf_heat_pump_suitability/research/exploratory/` - I dont think you'll need a subdir there :)

Please could you add a summary description of the notebook at the top. Can be just a few bullet points about what it contains.

crispy-wonton commented on asf_heat_pump_suitability/pipeline/run/heat_network_city_centres.py:
```diff
@@ -78,7 +80,45 @@ def parse_arguments() -> argparse.Namespace:
+        )
+
+        # Label UPRNs in city centres
+        city_centre_types = [  # TODO confirm types
```
```suggestion
        city_centre_types = [  # TODO confirm types with scaling
```
I think this is especially important!

eglucas committed:
edit spatial signature data getter docstring

crispy-wonton reviewed (changes requested):
Thanks for this @eglucas ! A great piece of analysis, great plotting, and thoroughly thought through - great work :D 

I've left some comments and suggestions. The main requested change is to make sure that we're addressing any edge cases of UPRNs that may arise when scaling which end up on the boundary between two spatially signatures meaning a single UPRN could receive 2 conflicting labels. You will find my comment about this in the `heat_network_zones.py` file.

To answer your qs in the PR description:

> Please inspect the new city_centre_uprn_df that is saved out to S3

I checked a few things - number of UPRNs is correct; proportions in city centre looks reasonable; no nulls in city centre label; spatial signatures are appropriately categorised into city centre and not. 

The only weird thing is there are some UPRNs which get `null` values for spatial signatures. I saw in your visualisations that there seems to be some gaps in the signatures, and can see you ultimately categorise these as not in city centres. I think it would be useful to note that somewhere as we might want to revisit it - especially when scaling, if there are larger gaps.

E.g. if there's an area with no spatial signature but is entirely enclosed by city centre, then maybe it should be classed as a city centre. Or maybe we should use a density metric in these areas instead? Not for this PR but good to think about if you have ideas.

> Whether you're happy with the saved location of the spatial signatures dataset, as well as how their paths are inserted in config

Paths of outputs and signatures datasets look good - I would just suggest combining city centre and HNZ output dfs into one file :)

> Schema of the saved output dataframe

Looks good !

> With regards to the analysis notebook, please run all the cells and have a look at the outputs. Let me know if you have any other thoughts on how we're evaluating whether the city centre signature types do a good enough job at flagging potential heat network zones. Please also flag if there are any errors in the analysis or interpretation.

Looks good to me and I like your visualisations, really helped to understand the data.

sofiapinto commented on asf_heat_pump_suitability/pipeline/transform/heat_network_zones.py:
```diff
@@ -54,3 +57,64 @@ def label_gdf_heat_network_zone_uprns(
+) -> pl.DataFrame:
+    """
+    Labels UPRNs that are located within a given Spatial Signature type deemed to be potentially
+    suitable for a heat network.
```
Happy to be challenged on this, but rather than talking about suitability for a heat network at this point I'd say something like:
"Labels UPRNs that are located within a given Spatial Signature type identified as a city centre" or similar.

Same about the function name.

In general it's a little confusing if we talk about DESNZ's heat network zones and then separately areas which are potentially suitable for HN. DESNZ's HN zones are also "potential"/planned. So i'd use different wording for spatial signatures - we're using them to identify high density areas/city centres, etc.

eglucas replied:
Thanks Sofia, that makes complete sense. 

FYI @crispy-wonton as well - I think it's best to make a clearer distinction between the existing HNZ labeller and this one for city centres. For complete clarity, I've moved this transformation function to a new separate module `transform/city_centres.py`. I'll rename the function and rewrite the docstring so that it's only related to city centre labelling (without making the leap that it could be a potential HNZ).

crispy-wonton replied:
Great, I think that's a good idea. Thanks @eglucas

sofiapinto commented on asf_heat_pump_suitability/pipeline/transform/heat_network_zones.py:
```diff
@@ -54,3 +57,64 @@ def label_gdf_heat_network_zone_uprns(
+        types (list, optional): spatial signature types potentially suitable for a heat network. Default is None to use all spatial signature types.
+
+    Returns:
+        pl.DataFrame: input UPRNs labelled with spatial signature identifiers and booleans indicating potential for a heat network
```
same as above

sofiapinto commented on asf_heat_pump_suitability/research/exploratory/20251205_plymouth_investigation.py:
```diff
@@ -0,0 +1,530 @@
+    edgecolor="blue",
+    legend=True,
+    column="ZoneID",
+)  # Existing heat network zones
```
```suggestion
)  # Existing/planned heat network zones
```

sofiapinto commented on asf_heat_pump_suitability/analysis/spatial_signatures_hn_zones/20251205_plymouth_investigation.py:
```diff
@@ -0,0 +1,530 @@
+# %% [markdown]
+# ### **Investigating areas in Plymouth LA with existing heat network zones or potential for a heat network**
```
Or similar wording...
```suggestion
# ###**Investigating areas in Plymouth LA with existing/planned heat network zones or areas which are city centres**
```

sofiapinto commented on asf_heat_pump_suitability/research/exploratory/20251205_plymouth_investigation.py:
```diff
@@ -0,0 +1,530 @@
+    plymouth_hn_zones_gdf,
+    spatial_signatures_gb_simplified_gdf,
+]
+len({gdf.crs for gdf in gdfs}) == 1
```
smart :)

sofiapinto commented on asf_heat_pump_suitability/research/exploratory/20251205_plymouth_investigation.py:
lots of great work here! :)

sofiapinto replied:
i'd rephrase things using "city centre" rather than "potential HNZ" when talking about signatures data, as per comments left in other files.

sofiapinto commented on asf_heat_pump_suitability/research/exploratory/20251205_plymouth_investigation.py:
```diff
@@ -0,0 +1,530 @@
+
+# %% [markdown]
+# 05/12/25: Computing the spatial overlap, the **city centre types cover 35% of the area covered by existing heat network zones**. Conversely, the **existing heat network zones cover 83% of the area covered by city centre types**.
+# - 35% metric -> The selected signature types only partially align with the existing heat network coverage; most of the current HNZ lie outside "city centre" areas. This indicates that using "city centre" areas is not able to capture all areas that have been deemed suitable for a heat network in reality.
```
And that's fine! I think we never expected this to capture all areas in the planned HNZ. The idea of having both is exactly to capture all areas which either have plans for a HNZ (which we know is also based on cost and other feasibility factors) with areas which we know from this new source that are of high density/which also represent city centres. We're not trying to see if the spatial signatures is a good proxy for DESNZ's HNZ.

sofiapinto reviewed (commented):
So much great work @eglucas! Also love how quickly you learnt about all these datasets & using/visualising geo data.

I did a quick review and left comments mostly about how we phrase things on planned HNZ vs identifying city centres.

I quickly ran the notebook, but didn't review in detail. I also haven't run the main script, but hopefully @crispy-wonton has :)

eglucas committed:
improve getter docstring

eglucas committed:
refactor city centre labeller function

eglucas committed:
refactor city centres labeller

eglucas committed:
refactor script

eglucas committed:
move city centre labeller function

eglucas committed:
exploratory notebook fixes

eglucas commented:
Hi @crispy-wonton and @sofiapinto 

Thank you so much for your comments, super helpful!

I've addressed all your comments and made changes. To summarise the main changes:

`load_gdf_spatial_signatures_gb()`
- docstrings have been rewritten to more accurately reflect the difference between the "simplified" and the "full" geopackage file being loaded
- `detail_level = "full"` is now the default

Labeller transformation function has been renamed to `label_gdf_city_centre_spatial_signatures_uprns()` and is now in a new file `pipeline/transform/city_centres.py`
- City centre spatial signature types are defined as a global variable at the top of the module, and is the default for the `types` argument
- the boolean column is now named `"in_city_centre"
- To address the cases where multiple spatial signature types are matched to one UPRN: assigned spatial signature types are now compiled in a list for each UPRN under the column "spatial_signature_types"
- To address how UPRNs that do not get a spatial signature match and therefore do not get classified to be in a city centre: No code fix for now but I have put a TODO comment on line 78 as a reminder to revisit in future

All language used when using the city centre labeller function in the run script and in the exploratory notebook has been modified from referring to these areas as 'potential hn zones' to simply 'city centre areas'.

`run/heat_network_city_centres.py`
- `in_hn_zone` and `in_city_centre` boolean columns are added to the same dataframe and saved as a single output
- **NOTE: I still specify `detail_level="simplified" here because I can't test run the script otherwise. Feel free to change this in the final run script for all features.**

Exploratory notebook
- this has been moved to `research/exploratory`
- explainer markdown has been added at the top
- language modifications
- small code fixes and improvements

Please have a look and merge when/if you're happy!

crispy-wonton commented on asf_heat_pump_suitability/pipeline/transform/city_centres.py:
```diff
@@ -0,0 +1,88 @@
+    # Add city centre boolean label
+    labelled_uprn_gdf["in_city_centre"] = labelled_uprn_gdf["type"].apply(
+        lambda x: any(t in types for t in x)
+    )
```
```suggestion
    # Add city centre boolean label
    labelled_uprn_gdf["in_city_centre"] = labelled_uprn_gdf["type"].isin(types)
    labelled_uprn_gdf = labelled_uprn_gdf.groupby("UPRN", group_keys=False).agg(
        {
            **{col: "first" for col in uprn_columns},  # keep original UPRN columns
            "geometry": "first",  # keep the point geometry
            "type": list,  # combine types into a list
            "in_city_centre": sum,  # sums >0 indicate UPRN in a city centre signature
        }
    # Convert city centre to boolean label
    ).astype({"in_city_centre": bool})
```
Just thinking about scaling, we might not want to use `apply` because of speed constraints. I think this works but haven't tested.

eglucas replied:
Just tested it, and yes it works! Will make that change and push.

eglucas committed:
refactor city centre bool label

crispy-wonton committed:
update heat_network_city_centres.py script to use full res spatial signatures

crispy-wonton committed:
Merge branch '200-filter-uprns-in-city-centre' of github.com:nestauk/asf_heat_pump_suitability into 200-filter-uprns-in-city-centre
